pmd-core/src/main/java/net/sourceforge/pmd/autofix/RuleViolationFixesApplier.java

```diff
@@ -0,0 +1,66 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.autofix;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import net.sourceforge.pmd.lang.ast.Node;
+
+/**
+ * Stores and applies, in order of arrival, insert / modify / delete operations over AST nodes.
+ */
+public class RuleViolationFixesApplier {
+    private final List<RuleViolationFixContext> ruleViolationFixContexts = new LinkedList<>();
+    private boolean applied = false;
+
+    /**
+     * Add a new rule violation fix with its corresponding node.
+     *
+     * @param ruleViolationFixClass the class to use when attempting to fix the node
+     * @param nodeToFix the node on which to apply operations
+     */
+    public void addRuleViolationFix(final Class<? extends RuleViolationFix> ruleViolationFixClass, final Node nodeToFix) {
+        assertFixesHaveNotBeenApplied();
+        ruleViolationFixContexts.add(new RuleViolationFixContext(ruleViolationFixClass, nodeToFix));
+    }
+
+    /**
+     * Apply, in order of insertion, the auto fixes for the nodes.
+     */
+    public void applyAutoFixes() {
+        assertFixesHaveNotBeenApplied();
+        applied = true;
+        for (final RuleViolationFixContext ruleViolationFixContext : ruleViolationFixContexts) {
+            ruleViolationFixContext.applyFixToNode();
+        }
+    }
+
+    /**
+     * Apply, in order of insertion, the auto fixes for the nodes. In the end it calls the {@link #clear()} method.
+     */
+    public void applyAutoFixesAndClear() {
```

Comment:
what's the benefit in reusing the same instance? Why do we keep both versions of these methods? It seems to me we will always want to use the later, and that it's a hack just to avoid instantiating more `RuleViolationFixesApplier`. To what end, I'm not sure